File: Day3/day3.js
const inputLines = readLineByLine('input.txt');

let sumErrors = 0;
for (const line of inputLines) {
    sumErrors += getPriority(findError(line));
}

console.log(`[Part1] Sum of priorities: ${sumErrors}`);

let sumRecurring = 0;
for (let i = 0; i < inputLines.length; i += 3) {
    const recurring = findRecurringInGroup([
        inputLines[i],
        inputLines[i + 1],
        inputLines[i + 2],
    ]);

    sumRecurring += getPriority(recurring.charCodeAt(0));
}

console.log(`[Part2] Sum of all groups: ${sumRecurring}`);

function getPriority(letter) { return letter > 96 ? letter - 96 : letter - 38; }

function findError(compartment) {
    const occurence = new Set();
    const half = compartment.length / 2;

    for (let i = 0; i < compartment.length; i++) {
        const codeUnit = compartment.charCodeAt(i);
        if (i < half) {
            occurence.add(codeUnit);
        } else if (occurence.has(codeUnit)) {
            return codeUnit;
        }
    }

    throw "could not find any recurring letter";
}

function getUniqueAscii(text) {
    const uniqueLetters = new Set();
    for (const letter of text) {
        uniqueLetters.add(letter);
    }
    return uniqueLetters.values();
}

function findRecurringInGroup(group) {
    const occurences = new Map();

    for (const compartment of group) {
        const compartmentUnique = getUniqueAscii(compartment);
        for (const letter of compartmentUnique) {
            const occurence = occurences.get(letter);
            occurences.set(letter, (occurence ?? 0) + 1);
        }
    }

    return getThreeOccurenceLetter(occurences);
}

function getThreeOccurenceLetter(map) {
    for (const [key, value] of map.entries()) {
        if (value == 3) return key;
    }
    throw "could not find 3 times recurring letter in a group";
}

function readLineByLine(filePath) {
    const fs = require('fs');
    const allFileContents = fs.readFileSync(filePath, 'utf-8');
    const inputLines = allFileContents.split(/\r?\n/);

    return inputLines;
}

File: Day3/day3.rs
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    let file = File::open("input.txt").unwrap();
    let reader = BufReader::new(file);
    let lines: Vec<String> = reader.lines().collect::<Result<_, _>>().unwrap();

    let sum_errors = lines.iter().fold(0i32, |acc, line| {
        acc + get_priority(find_error(line)) as i32
    });

    println!("[Part1] Sum of priorities: {}", sum_errors);

    let mut sum_recurring = 0i32;
    for i in (0..lines.len()).step_by(3) {
        let recurring = find_recurring_in_group(vec![&lines[i], &lines[i + 1], &lines[i + 2]]);
        sum_recurring += get_priority(recurring) as i32;
    }
    println!("[Part2] Sum of all groups: {}", sum_recurring);
}

fn get_priority(letter: u8) -> u8 {
    if letter > 96 {
        letter - 96
    } else {
        letter - 38
    }
}

fn find_recurring_in_group(group: Vec<&str>) -> u8 {
    let mut occurences: HashMap<u8, i32> = HashMap::new();
    for compartment in group.iter() {
        let compartment_unique = get_unique_ascii(compartment.to_string());

        for letter in compartment_unique.into_iter() {
            occurences
                .entry(letter)
                .and_modify(|occurence| *occurence += 1)
                .or_insert(1);
        }
    }

    return get_three_occurence_letter(occurences);
}

fn get_three_occurence_letter(map: HashMap<u8, i32>) -> u8 {
    for (key, value) in map.into_iter() {
        if value == 3 {
            return key;
        }
    }
    panic!("could not find 3 times recurring letter in a group");
}

fn get_unique_ascii(text: String) -> HashSet<u8> {
    text.as_bytes()
        .iter()
        .fold(HashSet::new(), |mut prev, code| {
            prev.insert(*code);
            prev
        })
}

fn find_error(compartment: &str) -> u8 {
    let mut occurence: HashSet<u8> = HashSet::new();
    let ascii_list = compartment.as_bytes();
    let half = ascii_list.len() / 2;

    for i in 0..ascii_list.len() {
        let code_unit = ascii_list[i];
        if i < half {
            occurence.insert(code_unit);
        } else if occurence.contains(&code_unit) {
            return code_unit;
        }
    }

    panic!("could not find any recurring letter");
}
